### DISCUSSION TITLE
Small problem when trying to follow `Providing Liquidity` section

### DISCUSSION THREAD

**Discussion by f3d0ss** [Jeiwan/uniswapv3-book#30] [Milestone 1. First Swap] [answered]
The followings are "problems" or things that I think need some clarification in the book in the Providing Liquidity section:

- `IERC20` and `IUniswapV3MintCallback` interfaces are used but not defined, they can be found in the [github branch of milestone_1](https://github.com/Jeiwan/uniswapv3-code/tree/milestone_1)  but I think it may be good to talk about it before they are used, or at least put a link to those interface contract that need to be added to the foundry project.

- In the test file in `setupTestCase()`  the global variable `shouldTransferInCallback` is updated after the `mint` function,  this way if `shouldTransferInCallback` is initialized to false/not initialized the callback function do not send the tokens, maybe the `shouldTransferInCallback` should be updated before the `mint` function is called

Those 2  "problems" cause the test to fail

**Comment by Jeiwan** [comment, marked as answer]
> IERC20 and IUniswapV3MintCallback interfaces are used but not defined

Yeah, I omitted a lot of auxiliary code in the book for brevity. Each milestone introduction has a link to full code:
https://github.com/Jeiwan/uniswapv3-book/blob/main/content/docs/milestone_1/introduction.md?plain=1#L37
So, I assume that readers will always follow the links to find missing parts.

> In the test file in setupTestCase() the global variable shouldTransferInCallback is updated after the mint function

Explained here: https://github.com/Jeiwan/uniswapv3-book/issues/40 We don't want to fail in `setupTestCase`, but in `testMintInsufficientTokenBalance` and `testSwapInsufficientInputAmount`.

**Reply by f3d0ss** [reply]
Thanks for the response and for your work.

For the second problem I just realized that `shouldTransferInCallback` should be initialized to `true`, like  `transferInMintCallback` and `transferInSwapCallback` in the final version of the milestone_1 test file.
